### ISSUE TITLE
v.rast.stats: warning instead of an error when no categories are found in raster map

### ISSUE THREAD

**Issue by pesekon2** [OSGeo/grass#1257]
**Describe the issue**

When you run `v.rast.stats` on a vector that is not in your computational region and does not have an overlap with the queried raster, it raises an error saying that `No categories found in raster map`. However, the behaviour of other modules is not to do so and to finish (e.g., `v.what.rast`). I would expect the same, systematic behaviour, i.e. to populate the columns with `NULL` values.

However, as the `No categories found in raster map` message is definitely an interesting information, I propose to change the error to a warning. What do you think?

**To Reproduce**

In:

```
g.region n=1 s=0 w=0 e=1 nsres=1 ewres=1
r.random.surface out=rast
g.region n=2 s=1 w=0 e=1 nsres=1 ewres=1
v.random out=vect npoints=1
v.db.addtable vect
g.region n=1 s=0 w=0 e=1 nsres=1 ewres=1
v.rast.stats  map=vect raster=rast column_prefix=rast
```

Out:

```
Generating raster map <rast>...
 100%
Writing raster map <rast>...
 100%
r.random.surface complete.
Generating points...
 100%
Building topology for vector map <vect@PERMANENT>...
Registering primitives...
WARNING: Values in column <cat> will be overwritten
Reading features...
 100%
Updating database...
 100%
1 categories read from vector map (layer 1)
1 categories read from vector map don't exist in selection from table
1 records updated/inserted (layer 1)
WARNING: No areas selected from vector map <vect>
ERROR: No categories found in raster map
```

An example with `v.what.rast`:

In: 

```
g.region n=1 s=0 w=0 e=1 nsres=1 ewres=1
r.random.surface out=rast
g.region n=2 s=1 w=0 e=1 nsres=1 ewres=1
v.random out=vect npoints=1
v.db.addtable vect
g.region n=1 s=0 w=0 e=1 nsres=1 ewres=1
v.what.rast  map=vect raster=rast column=rast
```

Out:

```
r.random.surface out=rast
g.region n=2 s=1 w=0 e=1 nsres=1 ewres=1
v.random out=vect npoints=1
v.db.addtable vect
g.region n=1 s=0 w=0 e=1 nsres=1 ewres=1
v.what.rast  map=vect raster=rast column=rast
Generating raster map <rast>...
 100%
Writing raster map <rast>...
 100%
r.random.surface complete.
Generating points...
 100%
Building topology for vector map <vect@PERMANENT>...
Registering primitives...
WARNING: Values in column <cat> will be overwritten
Reading features...
 100%
Updating database...
 100%
1 categories read from vector map (layer 1)
1 categories read from vector map don't exist in selection from table
1 records updated/inserted (layer 1)
Column <rast> not found in the table <vect>. Creating...
Reading features from vector map...
No features of type (point) found in vector map <vect>
```

**Comment by ninsbl** [comment]
The reason to use an error message instead of a warning is (from my understanding) that all subsequent processes are pointless if NO geometry is rasterized. Continuing would lead to unnecessary processing (esp. r.univar).

While I do see your consitency argument and I can see that in scripted tasks this behavior can be a problem, I am not sure if just chaning error to warning (as in #1258) is the best course of action.

If you want the module to succed in that case, I would probably prefer that columns are created and the module stops right after that with 0 exit code...

**Comment by pesekon2** [comment]
A good point. For some reason, I thought that processes like `r.univar` are needed because they report at least the number of cells and the number of `NULL` cells (as `v.to.rast` actually exports raster with `NULL` values in all pixels). However, checking the output now, these columns are also filled with `NULL` values. I will update the pull request to skip those processes.

By the way, what do you think about the behaviour I have just described. Would it be desired as another issue/PR, or does it make more sense with these `NULL` values?

**Comment by ninsbl** [comment]
I think it should be OK, to alter the PR.
And it should be safe to skip r.univar and the like (after https://github.com/OSGeo/grass/blob/1b0fe06cea82e34d50e2c32018795664174f4f6b/scripts/v.rast.stats/v.rast.stats.py#L283 right?)

**Comment by pesekon2** [comment]
Yes, it looks like the right place. I will refactor the entire `main()` a bit to split it into logical segments in the form of function calls - besides the fact that it is more readable, it will also give us the possibility of skipping also `v.category`.

**Comment by pesekon2** [comment]
The PR was updated. Now, there is quite a lot of changes. In a nutshell:

* the `main()` function rewritten into smaller, (hopefully) logical functions
* DBF checks moved to precede the number of categories check (useful to create columns)
* in the case of no categories in the raster, a warning is raised, columns created with `NULL` values and the module exits
* some logs taken out of the loop through rasters, as they actually should not be inside

As you have - due to this issue - some awareness about that, I have tagged you as a reviewer. However, feel free to untag.
